pageserver/src/tenant/mgr.rs

```diff
@@ -279,24 +280,44 @@ pub async fn create_tenant(
     ctx: &RequestContext,
 ) -> Result<Arc<Tenant>, TenantMapInsertError> {
     tenant_map_insert(tenant_id, |vacant_entry| {
+        let succeed = Cell::new(false);
+
         // We're holding the tenants lock in write mode while doing local IO.
         // If this section ever becomes contentious, introduce a new `TenantState::Creating`
         // and do the work in that state.
         let tenant_directory = super::create_tenant_files(conf, tenant_conf, tenant_id, CreateTenantFilesMode::Create)?;
-        // TODO: tenant directory remains on disk if we bail out from here on.
-        //       See https://github.com/neondatabase/neon/issues/4233
+
+        scopeguard::defer! {
+            if !succeed.get() {
+                std::fs::remove_dir_all(&tenant_directory).ok(); // ignore the error
+            }
+        }
 
         let created_tenant =
             schedule_local_tenant_processing(conf, &tenant_directory, remote_storage, ctx)?;
-        // TODO: tenant object & its background loops remain, untracked in tenant map, if we fail here.
-        //      See https://github.com/neondatabase/neon/issues/4233
+
+        let tenant_cloned = Arc::clone(&created_tenant);
+        scopeguard::defer! {
+            if !succeed.get() {
+                // As we might have removed the directory, the tenant should directly go into the broken state.
+                tenant_cloned.set_broken("failed to create".into());
+            }
+        }
+
+        fail::fail_point!("tenant-create-fail", |_| {
+            anyhow::bail!("failpoint: tenant-create-fail");
+        });
 
         let crated_tenant_id = created_tenant.tenant_id();
         anyhow::ensure!(
-                tenant_id == crated_tenant_id,
-                "loaded created tenant has unexpected tenant id (expect {tenant_id} != actual {crated_tenant_id})",
-            );
+            tenant_id == crated_tenant_id,
+            "loaded created tenant has unexpected tenant id (expect {tenant_id} != actual {crated_tenant_id})",
+        );
+
         vacant_entry.insert(Arc::clone(&created_tenant));
+
+        succeed.set(true);
```

Comment:
ScopeGuard would also have a bindable `let cleanup = ScopeGuard::guard((), |_| { ... });` api and `ScopeGuard::into_inner(cleanup)` for cancelling the cleanup to avoid the use of `Cell` and added condition. But I guess scopeguards docs give this example.

This would be even more applicable with the `tenant_cloned`.

(I am very much for drop guards in plain rust, which make it obvious that the code runs in a drop fn.)